src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1290,16 +1291,21 @@ private static BoundNode Rewrite(ImmutableDictionary<BoundExpression, (Nullabili
             return rewrittenNode;
         }
 
+        private static bool CanSkipAnalysis(CSharpCompilation compilation)
+        {
+            return compilation.LanguageVersion < MessageID.IDS_FeatureNullableReferenceTypes.RequiredVersion() || !compilation.IsNullableAnalysisEnabled;
+        }
+
         internal static bool NeedsAnalysis(CSharpCompilation compilation)
         {
+            var canSkipAnalysis = CanSkipAnalysis(compilation);
 #if DEBUG
-            // Always run analysis in debug builds so that we can more reliably catch
-            // nullable regressions e.g. https://github.com/dotnet/roslyn/issues/40136
-            return true;
-#else
-            var canSkipAnalysis = compilation.LanguageVersion < MessageID.IDS_FeatureNullableReferenceTypes.RequiredVersion() || !compilation.ShouldRunNullableWalker;
-            return !canSkipAnalysis;
+            if (canSkipAnalysis)
+            {
+                return !compilation.IsNullableAnalysisExplicitlyDisabled;
```

Comment:
Feels odd to have this check here but no tinside of `CanSkipAnalysis`

Reply:
If we moved the `IsNullableAnalysisExplicitlyDisabled` check into `CanSkipAnalysis()`, it would be difficult to differentiate cases in the callers where we want to analyze but drop results. (See the other two uses of `CanSkipAnalysis()`.)

---
In reply to: [541212255](https://github.com/dotnet/roslyn/pull/49876#discussion_r541212255) [](ancestors = 541212255)